### ISSUE TITLE
Printers need paper (not networks)

### ISSUE THREAD

**Issue by tlhackque** [simh/simh#69]
Some previous comments suggested making LPT devices into network printer clients.  This isn't a good idea.  

The simulated OSs that know how to print to network printers can do so via the network.  These include VMS, TOPS-10/20 and others.

Those that don't expect paper sizes, forms and behaviors that modern network printers don't provide.  Trying to keep up with that for individual network printers would be a never-ending effort.

On the other had, it certainly is true that "modern improvements" don't make it easy to render ASCII files as intended by the simulated OS.  A solution is indicated.

The alternative is to teach connect the printer emulations to a PAPER emulator, allowing them to print on the forms and paper they expect.  The output of the paper emulation is Adobe Portable Document format - .pdf.  This is universally available, and is an official format for archiving documents.  Thus, I expect that the drivers and applications for modern printers will be able to render PDF files as intended for the foreseeable future.  And it isolates SimH from the ever-changing rendering/printing landscape.

This has been implemented and I expect that it will be coming to SimH sometime soon.  

Here is a sample (low-resolution screen capture, reduced for this forum):

![lpt](https://f.cloud.github.com/assets/332018/863529/86919fe4-f60b-11e2-8aef-243259074184.PNG)

Besides variations of the greenbar form shown, it is possible to image any custom form by scanning/converting it to a suitable .JPG.

The preliminary release implements this capability in two ways:

1) A stand-alone utility will take the ASCII output of current printer emulators and convert it to .PDF

2) There is a simple API that allows printer emulators to write .PDF directly, via

```
ATTACH <device> file.pdf
```

It allows for specifying desired form sizes and other characteristics.  e.g.

```
ATTACH <device> form=bluebar width=22cm columns=120 myfile.pdf
```

Current ASCII output is still available.

The preliminary release implements this for the PDP-11, VAX (pdp11_lp) and PDP-10 (pdp10_lp20) emulators.

The reason for listing this as an issue is so that other emulations are aware of the capability and adopt the technology.

I have reviewed all the current printers, but only changed those I could test.

The changes for most emulators are very simple:  the pdp11_lp change was adding two lines of C code, and two lines of help.

Emulators that have attach & detach routines, and do file IO where the UNIT is pointed to by "uptr" need only those two lines.  Others need a bit more work - I estimate 10 minutes/emulator of coding, plus testing.

The API provides additional functions, such as checkpointing files when a printer goes idle.  These are recommended, but not required.  pdp10_lp20 demonstrates many of these.

When released, look for the user documentation in simh_doc.doc.  The technical documentation is in sim_pdflpt.h.

I would be happy to assist with updating other emulations if their owners approve and someone with the resources will test the changes.

In the meantime, the utility - lpt2pdf - will provide a 2-step process.

**Comment by JonathanBelanger** [comment]
I like this functionality, and there may be a resolution to being able to cause a PDF file to be printed to a real printer (not necessarily an LPT).

My idea is as follows:  The PDF file name should have a count field within it (e.g., filename01.pdf, where the 01 is the count field).  When the idle detection is triggered, the file is closed (detached) and a new file is opened with the next counter (e.g. filename02.pdf) and attached to LPT.  Outside of SimH, a watcher can detect when the file is closed and print it to a real printer.

Sorry about such a late comment.  I only just started playing with SimH (want to get the Alpha emulator working - already have the ROM images).

~Jon.

**Comment by thorhalbert** [comment]
This has been one of my holy-grail functionalities that I've never figured
out how to do.
I've thought of setting up LAT printers and never made that work.

I have recently thought to create named fifo-pipe and have the LPT device
open that and have an external program do the LPT formatting and create PDF
files automatically.

---Thor

**Comment by markpizz** [comment]
Before this functionality can be merged into master branch of the github.com/simh/simh repo someone needs to agree to own maintenance of this functionality and to fix the outstanding issues that have already been identified.

**Comment by JonathanBelanger** [comment]
I agree Mark, but I'm only just getting into this and am not sure of my availability for the long term (though I am playing with some performance improvements in the VAX and Alpha code that I'm hoping may be significant - if not, then I will not request a merging into the master branch).  I did think of a potential issue with my suggestion above.  The current functionality, without doing a close/open, is unaffected by a slow system.  If the close/open is implemented as defined above, it would be relatively easy to have a print job get interrupted for the idle time period and incorrectly assume that there are 2 separate jobs.  Additionally, multiple jobs in the queue may not have enough delay between them to close/open properly.  I'm starting to think this is starting to open a can of worms.

I could rewrite the print symbiont to indicate an end of job (I worked with a guy in DEC, dave porter, who for April Fools Day in 1986, updated the print symbiont to print backwards).

Oh well, back to the drawing board.

~Jon.

**Comment by leobru** [comment]
An oft-desired feature of line printer emulation, adding a layer of nostalgic authenticity,  is the ability to imitate wobbly lines (drum printers) or shifted characters (chain/train/band/bar printers) due to varying latency in serving printer interrupts. This will require a whole new API, of course.
